Subsample dataset by volume fix

z-fabian opened the pull request:
If `sample_rate < 1.0`, `SliceDataset` subsamples the training dataset by slices and not by volumes. This is undesired since 
1. Based on the function documentation says `sample_rate` should control what fraction of the volumes should be loaded. 
2. Subsampling by volumes makes more sense in practice. Randomly selected slices can create very unbalanced datasets (possibly lots of empty slices) and subsampling by volumes mimics the real life case of having less MRI volumes for training.  

This fix modifies `SliceDataset` to subsample by volumes when `sample_rate < 1.0`. If subsampling by slice is the desired behavior, then we should update the fuction helper to reflect that.

mmuckley reviewed (changes requested):
My worry here is we're breaking backwards compatibility. If someone has a project that depends on the repository and they want to reproduce an old result, this may break their code. Instead I would recommend a new variable, such as `volume_sample_rate`, and make sure that `sample_rate` and `volume_sample_rate` aren't set simultaneously, throwing an exception if they are. This might be best if both defaulted to `None` rather than `1.0`, but that's open to the implementation as you wish..

z-fabian commented:
@mmuckley As you suggested I added a new argument `volume_sample_rate`. If it is not set, `sample_rate` will behave the same way as before. If it is set, but `sample_rate` is not, then we subsample by volumes. I changed the default values for both of these arguments to None, in which case `1.0` will be used. Should be 100% backwards compatible.

mmuckley commented on fastmri/data/mri_data.py:
```diff
@@ -123,20 +124,40 @@ def __init__(
+        if sample_rates is not None and volume_sample_rates is not None:
+            raise ValueError(
+                "either set sample_rates (sample by slices) or volume_sample_rates (sample by volumes) but not both"
+            )
```
:+1:

mmuckley commented on fastmri/data/mri_data.py:
```diff
@@ -205,6 +234,11 @@ def __init__(
+        if sample_rate is not None and volume_sample_rate is not None:
+            raise ValueError(
+                "either set sample_rate (sample by slices) or volume_sample_rate (sample by volumes) but not both"
+            )
```
👍🏻

mmuckley commented on fastmri/pl_modules/data_module.py:
```diff
@@ -185,10 +206,12 @@ def prepare_data(self):
                 zip(data_paths, data_transforms)
             ):
                 sample_rate = self.sample_rate if i == 0 else 1.0
+                volume_sample_rate = self.volume_sample_rate if i == 0 else None
```
Funny enough, I think this is unnecessary.

mmuckley reviewed (approved):
Looks good to me! Great work!
